content/advent-2018/go-arrow.md

```diff
@@ -0,0 +1,531 @@
+
+## Lingua franca
+
+In [Data Science](https://en.wikipedia.org/wiki/Data_science) and in many scientific fields, the _lingua franca_ is Python.
+
+![xkcd-gravity](/postimages/advent-2018/go-arrow/xkcd-python.png)
+
+This means that a lot of the libraries are written in Python, with the CPU intensive parts leveraging either [NumPy](https://www.numpy.org/), [SciPy](https://www.scipy.org/) or, directly, a C/C++ library wrapped with the CPython C-API.
+This also means that:
+
+- the vast majority of the analysis pipelines are written in Python;
+- it is not completely straightforward to migrate parts of these pipelines to other languages, in an adiabatic and piecewise process.
+
+So why would anyone use Go for this?
+
+Well... the usual reasons:
+
+- Go is easier to deploy,
+- Go is faster than pure Python,
+- it is easier to write concurrent code in Go,
+- Go code and results are easy to reproduce,
+- Go code is more amenable to mechanical refactoring,
+- Go code tends to be more robust at scale than Python code.
+
+The real question is: do we have all the building blocks, all the libraries to write a modern, robust, efficient analysis pipeline in Go?
+
+Most of the science-y things one might carry out with the Python scientific stack (NumPy/SciPy) can be performed with [Gonum](https://gonum.org):
+
+- linear algebra, matrices,
+- differentiation,
+- statistics, probability distributions,
+- integration, optimization,
+- network creation,
+- Fourier transformations,
+- plots.
+
+There are also a few packages that enable file-level interoperability with many useful data science oriented or scientific "file formats":
+
+- [CSV](https://godoc.org/encoding/csv)
+- [SQL](https://godoc.org/database/sql)
+- [Numpy data files](https://godoc.org/github.com/sbinet/npyio)
+- [HDF5](https://godoc.org/gonum.org/v1/hdf5)
+- [FITS](https://godoc.org/github.com/astrogo/fitsio)
+- [HDFS](https://godoc.org/github.com/colinmarc/hdfs)
+- [Parquet](https://godoc.org/github.com/xitongsys/parquet-go/parquet)
+- [ORC](https://godoc.org/github.com/scritchley/orc)
+- [Cassandra](https://godoc.org/github.com/gocql/gocql)
+- ...
+
+Many basic ingredients are already there, even if some are still not there, yet (_e.g._ [ODE](https://en.wikipedia.org/wiki/Ordinary_differential_equation) is not implemented in [Gonum](https://gonum.org).)
+
+But even if basic file-level interoperability is (somewhat) achieved, one still needs to implement readers, writers and converters from one file format to any another when integrating with a new analysis pipeline:
+
+![file formats](/postimages/advent-2018/go-arrow/arrow-copy.png)
+
+In many scenarii, this implies a large fraction of the computation is wasted on serialization and deserialization, reimplementing over and over the same features for converting from `format-1` to `format-n` or from `format-1` to `format-m`, _etc..._
```

Comment:
Similar to "_lingua franca_" above. 
According to the internet, "_scenarii_" is a "(nonstandard, rare)" form, so maybe it will make sense to emphasise it  :)

```suggestion
In many _scenarii_, this implies a large fraction of the computation is wasted on serialization and deserialization, reimplementing over and over the same features for converting from `format-1` to `format-n` or from `format-1` to `format-m`, _etc..._
```

Reply:
done (used `scenarios`.)